Texture formats

swehner opened the pull request:
Hi,

These'd the changes to support the different formats - right now it chooses the format depending on the array's format - this might not be optimal (and the code still needs some error checking), but it should work both for loading a PIL image or an array directly.
All formats have a different number of channels except GL_ALPHA and GL_LUMINANCE which can be chosen using i_format.

BTW I've looked at the mipmapping part - and as it was right now memory was allocated for mipmaps even though the mipmap param was false (the filter method was set *after* loading the texture).
I'd also like to add a separate param to choose between GL_NEAREST and GL_LINEAR independent of if mipmaps are used or not (but that's for another PR).

That's the example I'm using: https://github.com/swehner/pi3d_demos/blob/test-texture-formats/TFormats.py

Memory usages on RaspberryPi with `/opt/vc/bin/vcdbg reloc` before the change (only 4 textures because the one with i_format=GL_ALPHA fails to load):
```
[  34] 0x1d324360: used 1.3M (refcount 2 lock count 1, size  1400832, align 4096, data 0x1d325000, d1rual) 'Texture blob'
[  50] 0x1d47b380: used 1.3M (refcount 2 lock count 1, size  1400832, align 4096, data 0x1d47c000, d1rual) 'Texture blob'
[  37] 0x1d5d23a0: used 1.3M (refcount 2 lock count 1, size  1400832, align 4096, data 0x1d5d3000, d1rual) 'Texture blob'
[  31] 0x1d7293c0: used 1.3M (refcount 2 lock count 1, size  1400832, align 4096, data 0x1d72a000, d1rual) 'Texture blob'
```
Memory usages on RaspberryPi with the changes:
```
[  48] 0x1d57b340: used 516K (refcount 2 lock count 1, size   524288, align 4096, data 0x1d57c000, d1rual) 'Texture blob'
[  45] 0x1d5fc360: used 1.0M (refcount 2 lock count 1, size  1048576, align 4096, data 0x1d5fd000, d1rual) 'Texture blob'
[  36] 0x1d6fd380: used 260K (refcount 2 lock count 1, size   262144, align 4096, data 0x1d6fe000, d1rual) 'Texture blob'
[  32] 0x1d73e3a0: used 1.0M (refcount 2 lock count 1, size  1048576, align 4096, data 0x1d73f000, d1rual) 'Texture blob'
[  81] 0x1d83f3c0: used 260K (refcount 1 lock count 0, size   262144, align 4096, data 0x1d840000, d1rual) 'Texture blob'
```
As you can see the allocated memory is a lot smaller - (according to number of channels except for the rgb image - maybe some memory alignment thing?) and also becaus no space for mipmaps is reserved.

Let me know what you think!
Before merging I'd like to do some more error checking on the numpy array format.
Also do you think it's be with it to convert the image format if r_format is specifide - so e.g. if you set i_format to GL_LUMINANCE and feed in an RGBA image should it convert it to L or load it as GL_RGBA?

paddywwoof commented:
Hi, thanks for this. Stopping the mipmapping looks to be a very significant improvement much better than using GL_RGBA4 etc - well spotted.

I've merged your changes into my local develop branch and all the standard tests seem fine but I've not tried any i_format arguments. I would say that if the user adds an i_format then the Texture class should try to convert it: RGB and RGBA converted to greyscale and that used as luminance or alpha, for RGBA to LA the greyscale should be used for luminance and the alpha mapped directly. Might be a bit of a mess!

Presumably the shader sets the RGB values of the pixels to 0.0 if the texture sampler is A?  It would be nice if it used the material values from Buffer.unib[3:6] but the alpha from texture2D()

I will have to look through the various classes that attempt to use Texture.alpha and tidy them up as this property now seems to be surplus to requirements! Also, I think arr = np.array(rgba)[:,:,2:4] is quicker and easier to understand than np.delete().

Finally I see you've used the double underscores name mangling whereas I've used 'weak private' underscores. It probably doesn't  matter either way but Texture is inherited by a few other classes and the mangling makes it a bit harder to use super functions.

swehner commented:
Hi,

I've added your suggestions to the branch:
I agree with the image conversion - it *is* a bit tricky, but I think it turned out more or less readable.
I've tried the pi3d_demo Gui and added the default conversion to RGBA - this changes the behaviour for this demo because the gifs used as icons now actually have the transparent background. (Before P images were converted to RGB)

Regarding the shader - I don't understand you very well because I don't know too much about them.
But isn't that what it's doing? I guess I'm not getting something here :S

The array slicing you propose is a lot nicer! I did have to construct an array of this again.
Not doing this doesn't produce the same result - it seems that the values are still there when passing it to OpenGL and the texture ends up distorted - not sure if this actually removes the values or simply changes the iterator...
This code seems to run as fast (maybe a bit faster), but it definitely is more readable. 
And I've changed the double_underscores as you mentioned!
Thanks again.

swehner commented:
I've added the patch I mentioned in the other thread that let's you choose LINEAR or NEAREST interpolation independent of whether mipmap creation is enabled.
When not specifying any value for filter it should behave exactly the same way as before.

paddywwoof commented:
@swehner Yes that's really annoying and I'm not sure why it (numpy) does it. I've seen similar before and it's due to the type changing. You can do this
arr = np.array(rgba)[:,:,2:4].astype(np.uint8)
I will merge your latest changes and have a play with it.

The shaders are divided into two types uv_... which look up the pixel values from a texture and mat_... which use a material RBG set in the Buffer (both then adjust for normal, light, reflections etc) As they stand the uv_ shaders don't take any notice of the material values, however it would be quite easy to alter. Possibly they should add the material RGB values, or add 2 * (material - [0.5, 0.5, 0.5]) (the default value).

swehner commented:
Ok, awesome! I tried to force dtype to uint8, but it didn't do anything... and it acutally reports dtype as being uint8...

I'll take a look at the shaders and see if I can learn more about them!

paddywwoof commented:
Excellent. Was there a reason you excluded LA 
... 
  if im.mode != expected_mode and self.i_format != GL_LUMINANCE_ALPHA:
seems to work ok with the second condition commented out and that allows RGB and RGBA images to be coerced to greyscale (if there's ever a requirement to do that!)

swehner commented:
Yeah, you're totally right - that isn't correct.
My idea was to avoid doing 2 conversions: RGBA to LA and then to get the  numpy array LA to RGBA.
But not doing them doesn't convert them to grayscale XD...
Thanks :)

paddywwoof commented:
Nice. I will add a version of that demo including mipmap=False, moving the Sprites backwards and forwards and changing the material colours. I have added a line to pi3d/shaders/std_main_uv.inc to allow 'tinting' with all the uv_ shaders
    texc.rgb += unib[1] - vec3(0.5);
I will check how much it effects frame rate but I suspect negligible, then push to origin develop.

swehner commented:
Ok - awesome!

paddywwoof commented:
@swehner I've merged this to develop but will hold off bumping master to v2.10 while there's a chance that @John-P will help get pi3d running on OS X

swehner commented:
Sure - no problem. Let me know if there's anything else that needs to be done for this!
